### PR TITLE
Kernel+NetworkSettings+Applications: Add a new Network Settings app

### PR DESCRIPTION
**Pull request by sppmacd**
This PR adds a basic NetworkSettings app. It supports only static IP assignment. The configuration is not permanent as it would require adding a separate service that would set settings on startup. The app is also unusable when run from Settings because it requires root to change interface settings.
image
It also includes LibGUI crash fix I noticed while writing GML for this.

### CODE REVIEW

**Comment by boricj** [comment]
Not directly related to this PR I guess, but do we actually need/want to support non-contiguous IPv4 subnet masks? If not, maybe we should consider changing the subnet mask field to another representation, such as `10.0.2.15/24`.

**Comment by sppmacd** [comment]
> Not directly related to this PR I guess, but do we actually need/want to support non-contiguous IPv4 subnet masks? If not, maybe we should consider changing the subnet mask field to another representation, such as `10.0.2.15/24`.

I chose this representation because ifconfig uses that. It is indeed a good idea, though.

**Comment by sppmacd** [comment]
New version:
image

**Comment by archanox** [comment]
I'm more partial to the original subnet dialog. It's more standard with practically all other network config dialogs out there. Whilst it's quite rare to see the second implementation in a GUI out in the wild.

Just my 2¢

**Comment by boricj** [comment]
Just because SerenityOS happens to be inspired by '90s user interfaces doesn't mean they can't be improved upon. Unless you are passing a Cisco professional certification, no sane person would ever configure non-contiguous IPv4 network masks anyway. CIDR is a practical representation used by network admins, whereas netmasks in decimal are error-prone and not worth the hassle, *especially* if one does not use networks sizes of 8, 16 or 24 bits long.

Also, IPv6 uses CIDR exclusively, so we might as well align to the sane convention.

**Comment by sppmacd** [comment]
With the CIDR notation it is much easier to calculate subnets. You can immediately see how much hosts will belong to the network, which cannot be done with the decimal notation. Also, as @boricj says, we don't need to stick to the more popular convention if the other is better.

**Comment by sppmacd** [comment]
I realized that it is broken because of introduction of routing table. Marking it as draft for now until #13857 is merged

**Comment by AtkinsSJ** [comment]
#13857 got merged, can this be un-drafted now?

**Comment by sppmacd** [comment]
Yes, but default gateway setting won't work because of route changes.

**Comment by AtkinsSJ** [comment]
Sorry this got ignored! I don't know if you're still interested in getting it merged, but there are a couple of issues:
- Merge conflict. (I think you can just drop the first commit, since the layout system has changed.)
- The "Apply" button never gets re-enabled. You'll need to call `set_modified(true);` when the user changes a setting, and make sure that this isn't triggered when changing those values in code. (A lot of Widget methods let you pass `AllowCallback:No` for this.)

Also any time I tried to change the settings and click "OK", I just got errors, so I'm not sure how useful this is in its current state. (Or maybe I'm using it wrong, networking isn't my thing. :sweat_smile: )

**Comment by sppmacd** [comment]
> Also any time I tried to change the settings and click "OK", I just got errors, so I'm not sure how useful this is in its current state. (Or maybe I'm using it wrong, networking isn't my thing. sweat_smile )

It works for me, but must be run as root. Default gateway setting doesn't work because of routing changes. I need to port it to work with NetworkServer anyway, so the PR is not ready to merge in the current state.

**Comment by sppmacd** [comment]
:ciflakes" 😢

**Comment by AtkinsSJ** [comment]
I shouldn't have taken so long to re-test this. Now I'm getting a crash just trying to launch it (either as root, or not):

```
11.902 NetworkSettings(44:44): ASSERTION FAILED: i < m_size
././AK/Vector.h:140

45.391 CrashReporter(51:52): --- Backtrace for thread #0 (TID 50) ---
45.391 CrashReporter(51:52): 0x0000000ba914602b: [/usr/lib/libsystem.so] syscall2 +0xb (syscall.cpp:25 => syscall.cpp:24)
45.391 CrashReporter(51:52): 0x00000016de07dbc6: [/usr/lib/libc.so] abort +0x26 (stdlib.cpp:223)
45.396 CrashReporter(51:52): 0x00000016de08376e: [/usr/lib/libc.so] __assertion_failed +0x7e (assert.cpp:33)
45.396 CrashReporter(51:52): 0x0000000b6483ad6b: [/bin/NetworkSettings] GUI::ItemListModel<AK::String, AK::Vector<AK::String, 0ul>, void>::data(GUI::ModelIndex const&, GUI::ModelRole) const +0x8b (Vector.h:140 => ItemListModel.h:77)
45.396 CrashReporter(51:52): 0x0000000a83c045df: [/usr/lib/libgui.so.serenity] GUI::ModelIndex::data(GUI::ModelRole) const +0x3f (ModelIndex.cpp:19)
45.396 CrashReporter(51:52): 0x0000000a83b99fae: [/usr/lib/libgui.so.serenity] GUI::ComboBox::selection_updated(GUI::ModelIndex const&) [clone .localalias] +0x7e (ComboBox.cpp:188)
45.400 CrashReporter(51:52): 0x0000000a83b9ada8: [/usr/lib/libgui.so.serenity] GUI::ComboBox::set_selected_index(unsigned long, GUI::AllowCallback) +0xe8 (ComboBox.cpp:227)
45.400 CrashReporter(51:52): 0x0000000b6483a1e0: [/bin/NetworkSettings] NetworkSettings::NetworkSettingsWidget::NetworkSettingsWidget() +0x1370 (NetworkSettingsWidget.cpp:92)
45.400 CrashReporter(51:52): 0x0000000b64836f3e: [/bin/NetworkSettings] AK::ErrorOr<AK::NonnullRefPtr<NetworkSettings::NetworkSettingsWidget>, AK::Error> GUI::TabWidget::try_add_tab<NetworkSettings::NetworkSettingsWidget>(AK::String) +0x4e (NetworkSettingsWidget.h:16 => TabWidget.h:61)
45.404 CrashReporter(51:52): 0x0000000b64837a63: [/bin/NetworkSettings] AK::ErrorOr<AK::NonnullRefPtr<NetworkSettings::NetworkSettingsWidget>, AK::Error> GUI::SettingsWindow::add_tab<NetworkSettings::NetworkSettingsWidget>(AK::String, AK::StringView) +0x63 (SettingsWindow.h:52)
45.404 CrashReporter(51:52): 0x0000000b64836bb3: [/bin/NetworkSettings] serenity_main(Main::Arguments) +0x283 (main.cpp:34)
45.404 CrashReporter(51:52): 0x0000000b648365ca: [/bin/NetworkSettings] main +0x9a (Main.cpp:39)
45.404 CrashReporter(51:52): 0x0000000b6483680d: [/bin/NetworkSettings] _entry +0x7d (crt0.cpp:54)
```

**Comment by sppmacd** [comment]
Should be working now (I was checking only interfaces added in config, not interfaces that actually exist)

**Comment by AtkinsSJ** [comment]
Hmmm. That's better, but still not quite right. I disabled both adapters, restarted Serenity, and now trying to launch NetworkSettings has that same crash.

Also, it's worth adding some kind of messages somewhere that 1) you need to be root to change things, and 2) changes only take effect after rebooting.

**Comment by sppmacd** [comment]
Should be working now. Also I added error when you try to run NetworkSettings as non-root and reboot warning when you set Enabled=false.

**Review by AtkinsSJ — APPROVED** [review]
LGTM! Just going to wait for CI

File: Kernel/Net/Routing.cpp
Comment (IdanHo) [inline Kernel/Net/Routing.cpp:156]: Too late, but this seems leftover from debugging.

Comment (sppmacd) [reply Kernel/Net/Routing.cpp:156]: #14523

**Pull request merged**
